### ISSUE TITLE
Replace all indexing by UInt

### ISSUE THREAD

**Issue by altavir** [mipt-npm/kmath#430]

**Comment by CommanderTvis** [comment]
What is the rationale?

**Comment by altavir** [comment]
Otherwise, we need to check argument positivity. It is better to check it on the type level. And we can always add extensions to accept ints as well.

**Comment by CommanderTvis** [comment]
> Otherwise, we need to check argument positivity. It is better to check it on the type level. And we can always add extensions to accept ints as well.

These extensions will perform the positivity checks, won't they?

**Comment by altavir** [comment]
I am not sure if simple toUInt does that, but we can always add this check.

**Comment by CommanderTvis** [comment]
> I am not sure if simple toUInt does that, 

It certainly doesn't.
![image](https://user-images.githubusercontent.com/38042667/140737197-d53ef284-6808-4325-b739-55d95f5e646f.png)

> but we can always add this check.
Then it will tremendous and invasive API change that is also related to huge consistency loss caused just by avoiding some checks? I think there should be some datum on performance gain we'll receive.

**Comment by altavir** [comment]
Then we will add safety checks

**Comment by alexismanin** [comment]
When talking about numeric indexing, is it limited to in-memory structures ?

Because if it is, as JVM arrays/collections are limited to Int. Requiring UInt would mean add overflow checks anyway.

In a larger scope, in-memory or not, I think you'll find more and more frameworks that use Long index instead of Int, to be able to represent datasets/structures with more than 2 (or 4) billion samples.

I work with spatial information, and there's often datasets that require long-based indices (one of the oldest example that come to my mind is the [BigTiff](http://bigtiff.org/) image format). But I think that apply even more on other domains like system observabillity, physics, etc.

To summarize my toughts : Modeling numerical indices is not such a trivial task, and depending of their aim (general-purpose indices, or in-memory purpose indices ?), multiple solutions should be evaluated (and each has its own complexity/drawbacks to deal with).

**Comment by altavir** [comment]
@alexismanin thanks for the comment. Indeed it is not that easy. In this case, we are talking only about non-negativity. Both value overflow and the negative index will give the same result, so I don't think it is a problem.

As for long indices, I was thinking about it some time ago and I came to conclusion, that long indices are not needed frequently, it is much easier to introduce partitions, which would allow us to use two numbers as indices instead of one and would solve the problem, The only additional cost would be to check partition switch on borders, which is not that expensive.

**Comment by CommanderTvis** [comment]
> @alexismanin thanks for the comment. Indeed it is not that easy. In this case, we are talking only about non-negativity. Both value overflow and the negative index will give the same result, so I don't think it is a problem.
> 
> As for long indices, I was thinking about it some time ago and I came to conclusion, that long indices are not needed frequently, it is much easier to introduce partitions, which would allow us to use two numbers as indices instead of one and would solve the problem, The only additional cost would be to check partition switch on borders, which is not that expensive.

I think in the context of human perception much more clear is that index can't be -1, rather than index can't be 2147483648.

**Comment by CommanderTvis** [comment]
I mean, what code looks more suspicious:
```
intIndexedThing[-1]
```
vs
```
uintIndexedThing[2147483648]
````

**Comment by altavir** [comment]
According to Roman Elizarov, UInt is not meant for indexing. So this issue is closed for now.
